pkg/ottl/functions_test.go

```diff
@@ -1141,54 +1141,66 @@ func functionWithEnum(Enum) (ExprFunc[interface{}], error) {
 	}, nil
 }
 
-type reusableFunctionFactory[A any] struct {
-	name string
-	args A
-	fn   any
-}
+// func createFunction[K any](fCtx FunctionContext, oArgs Arguments) (ExprFunc[any], error) {
+// 	args, ok := oArgs.(A)
 
-func (f reusableFunctionFactory[A]) Name() string { return f.name }
+// 	if !ok {
+// 		return nil, fmt.Errorf("reusableFunctionFactory args must be of type %T", f.args)
+// 	}
 
-func (f reusableFunctionFactory[A]) CreateDefaultArguments() Arguments {
-	return f.args
-}
+// 	funcVal := reflect.ValueOf(f.fn)
 
-func (f reusableFunctionFactory[A]) CreateFunction(fCtx FunctionContext, oArgs Arguments) (ExprFunc[any], error) {
-	args, ok := oArgs.(A)
+// 	if funcVal.Kind() != reflect.Func {
+// 		return nil, fmt.Errorf("a non-function value was passed to ottl.CreateFactory")
+// 	}
 
-	if !ok {
-		return nil, fmt.Errorf("reusableFunctionFactory args must be of type %T", f.args)
-	}
+// 	argsVal := reflect.ValueOf(args).Elem()
+// 	fnArgs := make([]reflect.Value, argsVal.NumField())
 
-	funcVal := reflect.ValueOf(f.fn)
+// 	for i := 0; i < argsVal.NumField(); i++ {
+// 		fnArgs[i] = argsVal.Field(i)
+// 	}
 
-	if funcVal.Kind() != reflect.Func {
-		return nil, fmt.Errorf("a non-function value was passed to ottl.CreateFactory")
-	}
+// 	out := funcVal.Call(fnArgs)
 
-	argsVal := reflect.ValueOf(args).Elem()
-	fnArgs := make([]reflect.Value, argsVal.NumField())
+// 	if !out[1].IsNil() {
+// 		return out[0].Interface().(ExprFunc[any]), out[1].Interface().(error)
+// 	}
 
-	for i := 0; i < argsVal.NumField(); i++ {
-		fnArgs[i] = argsVal.Field(i)
-	}
+// 	return out[0].Interface().(ExprFunc[any]), nil
+// }
 
-	out := funcVal.Call(fnArgs)
+func createFactory[A any](name string, args A, fn any) Factory[any] {
+	createFunction := func(fCtx FunctionContext, oArgs Arguments) (ExprFunc[any], error) {
+		args, ok := oArgs.(A)
 
-	if !out[1].IsNil() {
-		return out[0].Interface().(ExprFunc[any]), out[1].Interface().(error)
-	}
+		if !ok {
+			return nil, fmt.Errorf("reusableFunctionFactory args must be of type %T", args)
+		}
 
-	return out[0].Interface().(ExprFunc[any]), nil
+		funcVal := reflect.ValueOf(fn)
 
-}
+		if funcVal.Kind() != reflect.Func {
+			return nil, fmt.Errorf("a non-function value was passed to ottl.CreateFactory")
+		}
 
-func createFactory[A any](name string, args A, fn any) Factory[any] {
-	return reusableFunctionFactory[A]{
-		name: name,
-		args: args,
-		fn:   fn,
+		argsVal := reflect.ValueOf(args).Elem()
+		fnArgs := make([]reflect.Value, argsVal.NumField())
+
+		for i := 0; i < argsVal.NumField(); i++ {
+			fnArgs[i] = argsVal.Field(i)
+		}
+
+		out := funcVal.Call(fnArgs)
+
+		if !out[1].IsNil() {
+			return out[0].Interface().(ExprFunc[any]), out[1].Interface().(error)
+		}
+
+		return out[0].Interface().(ExprFunc[any]), nil
 	}
+
+	return NewFactory(name, args, createFunction)
 }
 
 func defaultFunctionsForTests() map[string]Factory[any] {
```

Comment:
Do we have a test for nil arguments?

Reply:
Sorry, got caught up dealing with rebase conflicts and didn't get to push these. We should have good branch coverage in `newFunctionCall` now, let me know if you see any cases that are missing.